### PR TITLE
use protobufs

### PR DESCRIPTION
**Pull request by tobowers**
Ok - I think this is a pretty interesting twist... use protobufs underneath the hood for the internal representation of the hamt (not for the values being stored).

Before:
```
$ go test . -run ^TestSetGet -v -count=1
=== RUN   TestSetGet
Total size is: 8340530, size of keys+vals: 6600000, overhead: 1.26
&{4713 100000 map[1:41261 2:16450 3:8613]}
start flush
flush took:  121.143488ms
put of root node start
put of root node took:  1.645556ms  pointer len:  256
finds took:  838.30603ms
--- PASS: TestSetGet (3.68s)
PASS
ok  	github.com/quorumcontrol/go-hamt-ipld	3.702s
```

After:

```
$ go test . -run ^TestSetGet -v -count=1
=== RUN   TestSetGet
Total size is: 8004956, size of keys+vals: 6600000, overhead: 1.21
&{4713 100000 map[1:41261 2:16450 3:8613]}
start flush
flush took:  27.161911ms
put of root node start
put of root node took:  103.124µs  pointer len:  256
finds took:  353.177223ms
--- PASS: TestSetGet (2.05s)
PASS
```

### CODE REVIEW

**Comment by tobowers** [comment]
fwiw: I want to talk this out, not just a quick accept on this. I *think* this makes sense and I don't see a real downside except that it's "different" than the rest of the Dag. And... fwiw: the DagProto format is kind of deprecated at the moment.

**Comment by tobowers** [comment]
and community for good measure:

Before:
```
goos: darwin
goarch: amd64
pkg: github.com/quorumcontrol/community/hub/tupelocache
BenchmarkAddAddBlockRequest-12    	    5000	   1574492 ns/op	  782081 B/op	   11468 allocs/op
PASS
ok  	github.com/quorumcontrol/community/hub/tupelocache	11.655s
Success: Benchmarks passed.
```


After:
```
oos: darwin
goarch: amd64
pkg: github.com/quorumcontrol/community/hub/tupelocache
BenchmarkAddAddBlockRequest-12    	   10000	    419228 ns/op	  396874 B/op	     547 allocs/op
PASS
ok  	github.com/quorumcontrol/community/hub/tupelocache	11.174s
Success: Benchmarks passed.
```

**Comment by tobowers** [comment]
If we're going to use a hamt for consensus then we will need to have a much faster version that what's in master now.

File: hamt.go [hunk 1]
```diff
@@ -7,38 +7,85 @@ import (
...
+	if err := pbNode.Unmarshal(bits); err != nil {
+		return err
+	}
+	n.Bitfield = new(big.Int).SetBytes(pbNode.Bitfield)
+	n.Pointers = fromProtobufs(pbNode.Pointers)
+	return nil
+}
+
+func (n *Node) Reset() {
+	// n.pbNode = new(pb.Node)
```
Comment (zonotope) [inline hamt.go]: probably should be uncommented

**Review by zonotope — COMMENTED** [review]
I like the idea of this a lot, I don't see any technical issues with the code on the first pass, and you can't argue with the numbers. My main questions however are around merging these changes upstream. This fork has already diverged a good bit, and this patch adds even more fundamental changes. Is the plan to maintain this fork going forward, or do you want to eventually get the changes accepted?

**Comment by tobowers** [comment]
I don’t think these will ever get merged back. The base storage layer is too different already (I didn’t open a PR for those). Moving forward they are going to reimplement a hamt anyway in ipld-prime. They are also staunch cbor supporters

**Comment by whyrusleeping** [comment]
Just a heads up, i'm going to be redoing the serialization layer for this pretty soon. Early results for other structures being similarly optimized show a 5-10x improvement in marshal speed, and a 3-5x improvement in unmarshal speed.

**Comment by tobowers** [comment]
Oh awesome. Where can I see the previous work?
-- 
Topper Bowers

topper@toppingdesign.com
github.com/tobowers
linkedin.com/in/topperbowers

**Comment by whyrusleeping** [comment]
Here's one of them: https://github.com/ipfs/go-cid/pull/88

Have some other stuff I havent quite pushed, and this repo is high on my list (its serialization is a bottleneck for us too :grimacing: )

**Comment by tobowers** [comment]
cool - well merging this for now while we wait for upstream.

**Pull request merged**